provider_test.go

```diff
@@ -268,8 +268,8 @@ func TestParseProviderSource(t *testing.T) {
 		},
 		"registry.Terraform.io/HashiCorp/AWS": {
 			Provider{
-				Type:      "aws",
-				Namespace: "hashicorp",
+				Type:      "AWS",
+				Namespace: "HashiCorp",
```

Comment:
One of the benefits of this library, though maybe implied, rather than explicitly documented, was canonical representation, which in turn made comparison and sorting possible.

For example, `hashicorp/aws` today is treated as equal to `HashiCorp/AWS` when it comes to cache entries (the lock file) etc. It is also one of the reasons Terraform LS uses this library and why we extracted it.

My understanding is that the intention was merely to allow `-`, so I'm assuming there isn't a strong reason to introduce case sensitivity?

Reply:
As best as I was able to determine, the private registry in Terraform Cloud (unlike the public registry) does not consider `myorg/aws` and `myorg/AWS` to be the same address, which would mean automatic lowercasing is a wrong behavior. 

But unlike the underscores, that one was based on a reading of the code rather than an actual experiment. I didn't attempt publishing a provider with capital letters in its name (which, to be fair, would be a pretty perverse thing to do, and I'm not sure anyone would try it if they weren't attempting to break something). 

Org names with uppercase letters are a bit more of a pressing concern, but I didn't attempt that yet either. I could give that a shot, it's somewhat easier than re-publishing my test provider with a vandalized name.

I'm open to adding some basic string lowercasing to restore case-collapse, if you're confident that we still need it.

Reply:
Yeah on further thought, if capitalized org names are a problem in TFC, then the solution is to add case normalization to its registry's routing, not to change the semantics over in this library. We already know TFC enforces org name uniqueness across any combination of letter cases, and the top-level org route accounts for that; if we missed handling that somewhere deeper in the routing, we'll just fix it.